From 07cdc9e5606064d72d0e2c57396f4c187542bfe4 Mon Sep 17 00:00:00 2001
From: Fatih Uzunoglu <fuzun54@outlook.com>
Date: Fri, 16 Jan 2026 11:19:34 +0200
Subject: [PATCH 1/1] Add missing include

---
 src/vector/vrle.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/vector/vrle.cpp b/src/vector/vrle.cpp
index c7a520f..5c0cd36 100644
--- a/src/vector/vrle.cpp
+++ b/src/vector/vrle.cpp
@@ -28,6 +28,7 @@
 #include <cstdlib>
 #include <cstring>
 #include <vector>
+#include <limits>
 #include "vdebug.h"
 #include "vglobal.h"
 
-- 
2.52.0

